### PR TITLE
Local history recording

### PR DESCRIPTION
**Pull request by utaal**
This introduces the optional feature (disabled by default) `record-history`: it can be enabled with `--features record-history` at build time; when this flag is not set, none of the supporting code is compiled in the verus binary.

After a per-project opt-in, this records locally every execution of verus on that project, using the machinery built for `--record`.

This is initially for internal use, so it is only documented in internal docs.

Inspired by @ahuoguo's work on #689.

@jonhnet has been successfully running this branch (built with the recording feature) recently.

### CODE REVIEW

**Review by jaybosamiya — COMMENTED** [review]
Overall this looks quite good to me; I've added some (minor) inline comments, but major comments that require further discussion here:

1. Do we want to store things in the `.record-history` in the same directory as the current implementation does?  To me, I'd prefer a centralized location for all projects (the projects opt-in by having a `.verus-record-history` empty file, anywhere in the ancestor of the project being verified; and all projects' git repositories go into a centralized location (`$XDG_STATE_HOME/verus-record` or `$XDG_CACHE_HOME/verus-record` or similar)).  This way, auditing the full list of recorded projects (and clearing things out and such) is way easier, rather than a user needing to do a `find` query.
2. A bare repo is better from perspective of hdd space usage, but most users aren't likely familiar with bare repos, maybe it might be worth considering a normal non-bare repo for auditing purposes?  I don't think it is too crucial either way though; anyone wanting to audit could just get used to bare git repos anyways.
3. Currently, if someone (even once) forgets to compile Verus with `--features record-history`, there doesn't seem to be any indication of the recordings not happening currently (even if they've opted a project in).  It might be good to let the user know "this project is opted in via {filename}, but Verus was compiled without `--features record-history` and thus will not be recording history" or similar to remind users that they forgot.  Especially when regularly updating Verus, I am worried about folks accidentally forgetting to pass the `--features` flag often.
4. Adding to point 4, is it necessary to have this be an off-by-default feature flag?  Currently it seems like two-level opt-in, which seems a bit unnecessary imho.  This feature as a whole is doing only local storage (and even then, it is with project-level opt-in) so I think it might be reasonable for us to keep the feature enabled by default (and paranoid users can disable it if needed), while still keeping the project-level opt-ins, so that it is not actually enabled by default on anything, and still requires an opt-in, but it makes it _much_ less likely that a user accidentally compiles without it, especially after having _already_ opted in.

File: tools/vargo/src/main.rs [hunk 1]
```diff
@@ -744,8 +776,21 @@ fn run() -> Result<(), String> {
             };
 
             for p in packages {
+                // || y.as_str() == "-F"
+                // || y.as_str() == "--features"
```
[the remark this answers is not in the document]
Comment (utaal) [reply tools/vargo/src/main.rs:780]: It's leftover from before. Removed. Thanks!

**Comment by utaal** [comment]
Thank you for the review. I addressed all the inline items, and https://github.com/verus-lang/verus/pull/798/commits/829e98fe4a4d996822a89fb2acd4607d0871b946 addresses "3. [...] if someone (even once) forgets to compile Verus with --features record-history, [...].

**Comment by utaal** [comment]
> 1. Do we want to store things in the .record-history in the same directory as the current implementation does?

This has the advantage that the opt in directory is exactly the place where history gets recorded, so the user does not need to think about multiple paths. I also kind of prefer not placing data elsewhere on a user's filesystem without an explicit choice, and I'd rather not need an additional configuration flag for this. Placing the recording "near" the project being verified makes it more likely that it's treated similarly to the rest of the project it records, and doesn't get accidentally get backed up / copied / forgotten alongside other projects. The general expectation when running a compiler is that it will modify the filesystem only in the vicinity of the source files, and the current approach doesn't break that expectation.

> 2. A bare repo is better from perspective of hdd space usage, but most users aren't likely familiar with bare repos, maybe it might be worth considering a normal non-bare repo for auditing purposes? I don't think it is too crucial either way though; anyone wanting to audit could just get used to bare git repos anyways.

That's a good point, but I wanted to minimize the risk of the repo ending up in an unexpected state that the code cannot handle. Not having a checkout reduces the chances of that happening.

> 4. Adding to point 4, is it necessary to have this be an off-by-default feature flag? [...]

Two reasons to have it as an off-by-default feature flag at least for now:

* it significantly affects the main code path that runs verus and prints output; I don't want a minor bug in the feature to affect users that do not care about recordings -- as we build confidence that it's working well on various platforms, we can reconsider this;
* I think we should err on the side of caution and not even have the recording code path in the binary unless a user explicitly asks for it: I don't want a subtle bug to cause inadvertent recordings for users that have not opted in, and we'd need some form of warning of disclaimer for an opt-out to be meaningful (users that just update regularly would not notice that they have just pulled a version with the recording feature enabled).

**Comment by utaal** [comment]
In the interest of not letting it diverge, and to simplify the workflow of folks relying on it, I'm going to merge this. Feel free to comment further if you think there are still open issues.

**Pull request merged**